knowledge-base/grid-convert-descriptors-to-sql.md

```diff
@@ -0,0 +1,203 @@
+tags: grid, blazor, filter descriptors , sort descriptors, SQL query
+res_type: kb
+ticketid: 1666625, 1653361
+---
+
+## Environment
+
+<table>
+    <tbody>
+        <tr>
+            <td>Product</td>
+            <td>Grid for Blazor</td>
+        </tr>
+    </tbody>
+</table>
+
+## Description
+
+When using the Grid [`OnRead` event]({%slug grid-events%}#read-event) to execute SQL queries, I need to convert the Grid's filter and sort descriptors into SQL query statements. This way I can create SQL clauses for filtering and ordering items directly through SQL.
+
+This KB article also answers the following questions:
+- How can I convert Grid filters and sorters to SQL `WHERE` and `ORDER BY` clauses?
+- Is there a way to parse Grid filter and sort descriptors into SQL queries?
+- Can I use [`DataSourceRequest`]({%slug common-features-data-binding-onread%}#event-argument) to generate SQL query statements for filtering and sorting?
+
+## Solution
+
+To convert the Grid's filter and sort descriptors into SQL query statements, you need to manually construct the SQL query within the `OnRead` event handler by utilizing the `args.Request.Filters` and `args.Request.Sorts` objects. Although Telerik UI for Blazor does not provide a direct method to extract the SQL query from the `DataSourceRequest`, you can achieve this manually.
+
+The following steps outline how to achieve this:
+
+1. Add an `OnRead` event to your Grid and in the event handler, access the `args.Request.Filters` and `args.Request.Sorts` to construct your SQL query.
+1. Iterate through `args.Request.Filters` to construct the `WHERE` clause of your SQL query. Each filter in this collection will correspond to a column filter in the Grid.
+1. Iterate through `args.Request.Sorts` to build the `ORDER BY` clause of your SQL query. Each sort descriptor corresponds to a column sorting in the Grid.
+1. Form your complete SQL query and execute it against your database with the constructed `WHERE` and `ORDER BY` clauses.
+1. Assign the result of your SQL query to the Grid by setting `args.Data`.
+
+## Example
+
+Below is a simplified example demonstrating how to parse filter and sort descriptors. This example does not directly execute a SQL query but outlines how to construct the `WHERE` and `ORDER BY` clauses.
+
+```csharp
+@using System.Text
+@using Telerik.DataSource
+@using Telerik.DataSource.Extensions
+
+<TelerikGrid TItem="@MyItem"
+             OnRead="@ReadItems"
+             FilterMode="@GridFilterMode.FilterRow"
+             Sortable="true"
+             Pageable="true">
+    <GridColumns>
+        <GridColumn Field=@nameof(MyItem.ID) />
+        <GridColumn Field=@nameof(MyItem.Name) />
+        <GridColumn Field=@nameof(MyItem.Age) />
+    </GridColumns>
+</TelerikGrid>
+
+@code {
+    private List<MyItem> GridData { get; set; }
+
+    private string SqlQuery { get; set; }
+    private string FilterQuery { get; set; }
+    private string SortQuery { get; set; }
```

Comment:
In a real-world case, there will be no three separate strings, but one, which contains the whole SQL query. This is not Telerik-related task, but will look a lot better if the example constructs a single SQL query instead of two. Also, since the query is not user anywhere, you can output it on the page for the user to see.